src/Stratis.Features.FederatedPeg/CollateralChecker.cs

```diff
@@ -0,0 +1,195 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.EventBus;
+using Stratis.Bitcoin.Features.BlockStore.Controllers;
+using Stratis.Bitcoin.Features.PoA;
+using Stratis.Bitcoin.Features.PoA.Events;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Features.FederatedPeg
+{
+    /// <summary>Class that checks if federation members fulfill the collateral requirement.</summary>
+    public class CollateralChecker : IDisposable
+    {
+        private readonly BlockStoreClient blockStoreClient;
+
+        private readonly IFederationManager federationManager;
+
+        private readonly ISignals signals;
+
+        private readonly ILogger logger;
+
+        /// <summary>Protects access to <see cref="depositsByAddress"/>.</summary>
+
+        private readonly object locker;
+
+        private readonly CancellationTokenSource cancellationSource;
+
+        private SubscriptionToken memberAddedToken, memberKickedToken;
+
+        /// <summary>Amount of confirmations required for collateral.</summary>
+        private const int RequiredConfirmations = 1;
+
+        private const int CollateralUpdateIntervalSeconds = 20;
+
+        /// <summary>Deposits mapped by federation member.</summary>
+        /// <remarks>
+        /// Deposits are not updated if federation member doesn't have collateral requirement enabled.
+        /// All access should be protected by <see cref="locker"/>.
+        /// </remarks>
+        private Dictionary<string, Money> depositsByAddress;
+
+        private Task updateCollateralContinuouslyTask;
+
+        public CollateralChecker(ILoggerFactory loggerFactory, IHttpClientFactory httpClientFactory, FederationGatewaySettings settings,
+            IFederationManager federationManager, ISignals signals)
+        {
+            this.federationManager = federationManager;
+            this.signals = signals;
+
+            this.cancellationSource = new CancellationTokenSource();
+            this.locker = new object();
+            this.depositsByAddress = new Dictionary<string, Money>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.blockStoreClient = new BlockStoreClient(loggerFactory, httpClientFactory, settings.CounterChainApiPort);
```

Comment:
Any reason not to inject this? Will be easier to test too

Reply:
Yes. BlockStoreClient accepts port as an integer so it can't be on DI without changing the constructor. 

Also there is no singletones registrator for mainnet gateway. We have gateway which is for side and main and we have side addon on top. So without a refactoring the only option is to add the client to both main and side gateways.

Reply:
@noescape00 if that's the case then you need to remove `BlockStoreClient blockStoreClient` from the constructor.

Reply:
Yup, already removed